---
title: "Read coverage Sendai and Yamanaka"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library(here)
library(tidyverse)
library(powerjoin)
library(data.table)
library(synExtra)
library(ggcoverage)
library(Rsamtools)

synapser::synLogin()

syn <- synDownloader("~/data", .cache = TRUE)
```

```{r}
inputs <- c(
  sample_meta = "syn53061904",
  size_factors = "syn53061911"
)

input_files <- map(
  inputs, syn
)

sample_meta <- read_csv(input_files[["sample_meta"]])
size_factors <- read_csv(input_files[["size_factors"]])
```

```{r}
ensembl_gtf <- rtracklayer::import(
  here("index", "Homo_sapiens.GRCh38.110.gtf.gz")
)
```

Select samples to plot.

```{r}
samples_sendai <- sample_meta %>%
  filter(
    (study_name == "Buckberry Nature 2023" &
      treatment %in% c("NtP", "Na√Øve", "Primed", "Control ESC", "TNT")) |
      source_name == "5iLAF SSEA4- subpopulation UCLA1 hESCs"
  )

```

## Sendai

Find alignments for selected samples.

```{r}
sendai_region <- "pSeV-idSOX2-Pmut:1150-19976"

sendai_bam_meta <- samples_sendai %>%
  mutate(
    # Label Pastor samples correctly
    treatment = replace_na(treatment, "Control ESC"),
    bam_name = paste0(sample_accession, "_Aligned.out_sorted"),
    bam_path = file.path("star", paste0(bam_name, ".bam")),
    bam_obj = map(
      bam_path, BamFile
    ),
    n_mapped = map_int(
      bam_obj,
      \(x) idxstatsBam(x)$mapped %>%
        sum()
    )
  ) %>%
  power_left_join(
    size_factors,
    by = "sample_accession",
    check = check_specs(
      unmatched_keys_left = "warn",
      duplicate_keys_right = "warn"
    )
  )
all(file.exists(sendai_bam_meta$bam_path))
```

Compute coverage of the Sendai virus genome using ggcoverage.

```{r}
sendai_bam_meta_gg <- sendai_bam_meta %>%
    select(sample_accession, SampleName = bam_name, size_factor, n_mapped)

sendai_bam_tracks <- LoadTrackFile(
  sendai_bam_meta$bam_path,
  format = "bam",
  region = sendai_region,
  meta.info = sendai_bam_meta_gg,
  norm.method = "None",
  extend = 0,
  bin.size = 500,
  # bc.extra.para = "--effectiveGenomeSize 2913022398"
) %>%
  as_tibble() %>%
  mutate(
    score_raw = score,
    score = 1e6 * score / (size_factor * mean(n_mapped))
  )
```

Collapse start and end intervals to a single column for plotting.

```{r}
sendai_bam_coverage_long <- sendai_bam_tracks %>%
  pivot_longer(c(start, end), names_to = "pos_type", values_to = "pos")
  # mutate(
  #   treatment_grouped = case_when(
  #     treatment == "Control ESC" ~ paste("Control ESC", study_name),
  #     TRUE ~ treatment
  #   )
  # ) %>%
  # group_by(treatment_grouped) %>%
  # mutate(
  #   rep = factor(
  #     sample_accession, labels = seq_len(as.character(length(unique(cur_data()$sample_accession))))
  #   )
  # ) %>%
  # # Only keep two replicates
  # filter(
  #   rep %in% c("1", "2")
  # ) %>%
  # ungroup()

write_csv(
  sendai_bam_coverage_long,
  "results/sendai_bam_coverage_long.csv.gz"
)
```

## Yamanaka

Prepare Yamanaka gene CDS and UTR coordinates.

```{r}
yamanaka_factors <- c("KLF4", "MYC", "POU5F1", "SOX2")

yamanaka_gtf <- ensembl_gtf[
  ensembl_gtf$gene_name %in% yamanaka_factors &
    replace_na(str_detect(ensembl_gtf$tag, "MANE_Select"), FALSE)
]
```

Quantify coverage of Yamanaka genes using ggcoverage.

```{r}
yamanaka_regions <- reduce(
  split(yamanaka_gtf, yamanaka_gtf$gene_name)
)

yamanaka_bam_tracks <- bind_rows(
  map(
    yamanaka_regions,
    \(x) LoadTrackFile(
      sendai_bam_meta$bam_path,
      region = paste0(seqnames(x), ":", start(x), "-", end(x)),
      format = "bam",
      meta.info = sendai_bam_meta_gg,
      norm.method = "None",
      extend = 1000,
      bin.size = 100,
      # bc.extra.para = "--effectiveGenomeSize 2913022398"
    )
  ),
  .id = "gene_name"
) %>%
  as_tibble() %>%
  mutate(
    score_raw = score,
    score = 1e6 * score / (size_factor * mean(n_mapped))
  )

yamanaka_bam_tracks_long <- yamanaka_bam_tracks %>%
  group_by(sample_accession, gene_name) %>%
  mutate(
    score_norm = scales::rescale(score, to = c(0, 1))
  ) %>%
  ungroup() %>%
  pivot_longer(c(start, end), names_to = "pos_type", values_to = "pos")

write_csv(
  yamanaka_bam_tracks_long,
  "results/yamanaka_bam_coverage_long.csv.gz"
)
```

```{r}
synStoreMany(
  c(
    "results/sendai_bam_coverage_long.csv.gz",
    "results/yamanaka_bam_coverage_long.csv.gz"
  ),
  parentId = "syn53061909",
  used = unname(inputs),
  forceVersion = FALSE
)

```
