---
title: "Plot Salmon quants"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library(here)
library(tidyverse)
library(powerjoin)
library(data.table)
library(synExtra)
library(ggbeeswarm)

synapser::synLogin()

syn <- synDownloader("~/data", .cache = TRUE)
```

```{r}
inputs <- c(
  sample_meta = "syn53061904",
  salmon_quants = "syn53061910",
  salmon_gene_quants = "syn53061989"
)

input_files <- map(
  inputs, syn
)

```

## Metadata

```{r}
sample_meta <- read_csv(input_files[["sample_meta"]])
salmon_quants <- read_csv(input_files[["salmon_quants"]])
salmon_gene_quants <- read_csv(input_files[["salmon_gene_quants"]])

```

```{r}
sendai_counts <- salmon_quants %>%
  filter(str_detect(transcript_id, fixed("Sendai")))

sendai_counts_avg <- sendai_counts %>%
  filter(transcript_id != "Sendai_F_gene") %>%
  group_by(sample_accession) %>%
  summarize(
    tpm = mean(tpm),
    count = mean(count),
    .groups = "drop"
  )

sample_meta %>%
  left_join(
    sendai_counts_avg,
    by = "sample_accession"
  ) %>%
  pull(tpm) %>%
  signif(digits = 2) %>%
  clipr::write_clip()
```


```{r}
# library(AnnotationHub)
# ah = AnnotationHub()

ensembl_gtf <- rtracklayer::readGFF(
  here("index", "Homo_sapiens.GRCh38.110.gtf.gz"),
  columns = c("seqid", "source", "type"),
  tags = c("transcript_id", "gene_id", "gene_name", "gene_biotype")
) %>%
  filter(gene_biotype == "protein_coding") %>%
  as_tibble()

ensembl_gtf %>%
  filter(type == "gene") %>%
  drop_na(gene_name, gene_id) %>%
  distinct() %>%
  group_by(gene_name) %>%
  filter(n() > 1) %>%
  ungroup() %>%
  arrange(gene_name)

gene_id_gene_symbol_map <- ensembl_gtf %>%
  filter(type == "gene") %>%
  drop_na(gene_name, gene_id) %>%
  distinct(seqid, source, gene_id, gene_name)

transcript_gene_map <- ensembl_gtf %>%
  filter(type == "transcript") %>%
  distinct(gene_id, transcript_id)
```

```{r}
yamanaka_factors <- c("KLF4", "MYC", "POU5F1", "SOX2")

marker_genes <- tribble(
  ~gene_name, ~marker_group,
  "VIM", "fibroblast",
  "PDGFRA", "fibroblast",
  "FAP", "fibroblast",
  "AIFM2", "fibroblast",
  "SNAI2", "fibroblast",
  "KLF17", "naive",
  "TBX3", "naive",
  "NANOG", "pluripotency"
) %>%
  bind_rows(tibble(gene_name = yamanaka_factors, marker_group = "yamanaka"))

```

```{r}
salmon_gene_quants_symbol <- salmon_gene_quants %>%
  left_join(
    gene_id_gene_symbol_map,
    by = "gene_id"
  )
```


```{r}

p <- salmon_gene_quants_symbol %>%
  power_inner_join(
    marker_genes,
    by = "gene_name",
    check = check_specs(
      unmatched_keys_right = "warn",
      duplicate_keys_right = "warn"
    )
  ) %>%
  power_inner_join(
    sample_meta,
    by = "sample_accession",
    check = check_specs(
      unmatched_keys_left = "warn",
      unmatched_keys_right = "warn",
      duplicate_keys_right = "warn"
    )
  ) %>%
  mutate(
    is_zero = tpm == 0,
    tpm = tpm + 0.01
  ) %>%
  group_nest(marker_group) %>%
  mutate(
    p = map(
      data,
      \(df) ggplot(
        df,
        aes(
          x = treatment,
          y = tpm,
          fill = gene_name,
          shape = is_zero
        )
      ) +
        geom_quasirandom(color = alpha("white", 0), alpha = 0.5) +
        stat_summary(aes(color = gene_name, group = gene_name), fill = NA, geom = "point", shape = "-", size = 5, fun.data = mean_se) +
        scale_y_log10(labels = scales::trans_format("log10", scales::label_math(10^.x))) +
        scale_shape_manual(values = c(`FALSE` = 21, `TRUE` = 25), guide = "none") +
        # facet_grid(cols = vars(study_name), scales = "free") +
        ggh4x::facet_grid2(cols = vars(study_name), scales = "free", space = "free_x") +
        guides(color = "none", fill = guide_legend(override.aes = list(shape = 21))) +
        theme_bw() +
        theme(
          axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1)
        )
    )
  )

dir.create("plots")
ggsave(
  here("plots", "marker_genes_tpms_grid.pdf"),
  patchwork::wrap_plots(
    p$p %>%
      imap(
        \(x, i) if (i != length(p$p)) x + theme(
          axis.text.x = element_blank(), axis.title.x = element_blank()
        ) else x
      ),
    ncol = 1
  ),
  width = 10, height = 10
)

```


```{r}
time_course_samples <- tribble(
  ~treatment, ~time_point, ~time_point_plot, ~stage,
  "Fibroblast", "P8", "P8", "Fibroblast",
  "Fibroblast", "P20 + 3", "P20 + 3", "Fibroblast",
  "Fibroblast", "D3", "D3", "Fibroblast",
  "Fibroblast", "D7", "D7", "Fibroblast",
  "Primed", "D13", "D13_", "Primed",
  "Primed", "D21", "D21", "Primed",
  "Primed", "P3", "P3", "Primed",
  "Primed", "P10", "P10", "Primed",
  "Primed", "P17", "P17", "Primed",
  "Primed", "P18", "P18", "Primed",
  "5iLAF", "D13", "D13", "Transition",
  "NHSM", "D13", "D13", "Transition",
  "t2iLGoY", "D13", "D13", "Transition",
  "RSeT", "D13", "D13", "Transition",
  "Naïve", "P15", "P15", "Naïve"
) %>%
  mutate(
    across(time_point_plot, \(x) fct_inorder(x, ordered = TRUE)),
    medium = if_else(
      stage == "Transition",
      treatment,
      NA_character_
    )
  )

stage_markers <- time_course_samples %>%
  group_by(stage) %>%
  summarize(
    start = min(time_point_plot),
    end = max(time_point_plot),
    .groups = "drop"
  )

time_course_data <- sample_meta %>%
  filter(study_name %in% c("Buckberry Nature 2023", "Liu Nature 2020")) %>%
  inner_join(
    time_course_samples,
    by = c("treatment", "time_point")
  ) %>%
  inner_join(
    salmon_quants %>%
      filter(
        str_starts(transcript_id, fixed("Sendai")),
        transcript_id != "Sendai_F_gene"
      ) %>%
      group_by(sample_accession) %>%
      summarize(tpm = mean(tpm), .groups = "drop"),
    by = "sample_accession"
  )

line_set_data <- tribble(
  ~set, ~stage,
  "Primed", "Fibroblast",
  "Primed", "Primed",
  "Naïve", "Fibroblast",
  "Naïve", "Transition",
  "Naïve", "Naïve"
) %>%
  inner_join(
    time_course_data,
    by = "stage",
    relationship = "many-to-many"
  ) %>%
  group_by(set, stage, time_point_plot) %>%
  summarize(tpm = mean(tpm), .groups = "drop")

STAGE_Y_POS = Inf

p <- time_course_data %>%
  # filter(transcript_id == "Sendai_N_gene") %>%
  ggplot(
    aes(
      time_point_plot,
      tpm,
      color = medium
    )
  ) +
    scale_x_discrete(
      limits = unique(time_course_samples$time_point_plot),
      labels = \(x) str_replace(x, "_", "")
    ) +
    scale_color_hue(
      breaks = na.omit(unique(time_course_data$medium)),
      na.value = "black"
    ) +
    geom_rect(
      aes(
        xmin = stage(start, after_stat = xmin - .5),
        xmax = stage(end, after_stat = xmax + .5),
        fill = stage
      ),
      data = stage_markers,
      alpha = 0.3,
      ymin = -Inf, ymax = Inf,
      inherit.aes = FALSE,
      show.legend = FALSE
    ) +
    scale_fill_brewer(palette = "Pastel2") +
    geom_quasirandom(
      aes(shape = study_name),
      dodge.width = .7
    ) +
    geom_text(
      aes(
        xmin = start, xmax = end, x = after_stat(.5 * (xmin + xmax)), label = stage
      ),
      data = stage_markers,
      y = STAGE_Y_POS,
      size = 3,
      hjust = .5, vjust = 1,
      inherit.aes = FALSE
    ) +
    geom_line(
      aes(
        x = time_point_plot,
        y = tpm,
        group = set
      ),
      data = line_set_data,
      inherit.aes = FALSE,
      color = "black",
      alpha = 0.8
    ) +
    geom_segment(
      aes(
        x = stage(time_point_plot, after_stat = x - .1),
        xend = stage(time_point_plot, after_stat = x + .1),
        y = tpm, yend = tpm
      ),
      data = line_set_data,
      inherit.aes = FALSE,
      color = "black",
      alpha = 0.8
    ) +
    theme_minimal() +
    theme(plot.background = element_rect(fill = "white", color = NA)) +
    coord_cartesian(clip = "off") +
    labs(y = "Average TPM Sendai genes", x = NULL, shape = "Source", color = "Medium")

ggsave(
  here("plots", "sendai_time_course2.pdf"),
  p,
  width = 8.5, height = 3
)



p <- time_course_data %>%
  mutate(
    is_zero = tpm == 0,
    tpm = tpm + 0.005
  ) %>%
  # filter(transcript_id == "Sendai_N_gene") %>%
  ggplot(
    aes(
      time_point_plot,
      tpm,
      color = medium
    )
  ) +
    scale_x_discrete(
      limits = unique(time_course_samples$time_point_plot),
      labels = \(x) str_replace(x, "_", "")
    ) +
    scale_color_hue(
      breaks = na.omit(unique(time_course_data$medium)),
      na.value = "black"
    ) +
    geom_rect(
      aes(
        xmin = stage(start, after_stat = xmin - .5),
        xmax = stage(end, after_stat = xmax + .5),
        fill = stage
      ),
      data = stage_markers,
      alpha = 0.4,
      ymin = -Inf, ymax = Inf,
      inherit.aes = FALSE,
      show.legend = FALSE
    ) +
    scale_fill_brewer(palette = "Pastel2") +
    geom_quasirandom(
      aes(shape = study_name),
      dodge.width = .7
    ) +
    geom_text(
      aes(
        xmin = start, xmax = end, x = after_stat(.5 * (xmin + xmax)), label = stage
      ),
      data = stage_markers,
      y = STAGE_Y_POS,
      size = 3,
      hjust = .5, vjust = 1,
      inherit.aes = FALSE
    ) +
    geom_line(
      aes(
        x = time_point_plot,
        y = tpm,
        group = set
      ),
      data = line_set_data %>%
        mutate(
          tpm = tpm + 0.005
        ),
      inherit.aes = FALSE,
      color = "black",
      alpha = 0.8
    ) +
    geom_segment(
      aes(
        x = stage(time_point_plot, after_stat = x - .1),
        xend = stage(time_point_plot, after_stat = x + .1),
        y = tpm, yend = tpm
      ),
      data = line_set_data %>%
        mutate(
          tpm = tpm + 0.005
        ),
      inherit.aes = FALSE,
      color = "black",
      alpha = 0.8
    ) +
    geom_hline(
      yintercept = 0.005,
      color = "black",
      linetype = "dashed"
    ) +
    theme_minimal() +
    scale_y_log10(labels = scales::trans_format("log10", scales::label_math(10^.x))) +
    theme(plot.background = element_rect(fill = "white", color = NA)) +
    coord_cartesian(clip = "off") +
    labs(y = "Average TPM Sendai genes", x = NULL, shape = "Source", color = "Medium")

ggsave(
  here("plots", "sendai_time_course_log.pdf"),
  p,
  width = 8.5, height = 3
)

```

```{r}
yamanaka_factors <- c("KLF4", "MYC", "POU5F1", "SOX2")

yamanaka_transcripts <- ensembl_gtf %>%
  filter(
    type == "transcript",
    gene_name %in% yamanaka_factors
  ) %>%
  distinct(transcript_id, gene_name, gene_id)

```

```{r}
p <- salmon_tpm %>%
  inner_join(
    transcripts_of_interest,
    by = "transcript_id"
  ) %>%
  mutate(
    tpm = tpm + 0.01
  ) %>%
  ggplot(
    aes(
      x = transcript_id,
      y = tpm,
    )
  ) +
    geom_quasirandom() +
    scale_y_log10() +
    facet_wrap(~gene_name, scales = "free_x") +
    theme_bw() +
    theme(
      axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1)
    )

dir.create("plots")
ggsave(
  here("plots", "yamanaka_all_tpms.pdf"),
  p,
  width = 10, height = 6
)

```



```{r}
library(ggbeeswarm)

sendai_tpm <- salmon_tpm %>%
  filter(
    str_starts(transcript_id, fixed("Sendai"))
  )

p <- sendai_tpm %>%
  mutate(
    tpm = tpm + 0.01
  ) %>%
  ggplot(
    aes(
      x = transcript_id,
      y = tpm,
    )
  ) +
    geom_quasirandom() +
    scale_y_log10() +
    theme_bw() +
    theme(
      axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1)
    )

dir.create("plots")
ggsave(
  here("plots", "sendai_all_tpm_beeswarm.pdf"),
  p,
  width = 10, height = 6
)

```

```{r}

p <- sendai_tpm %>%
  inner_join(
    sample_meta %>%
      filter(
        study_name == "Buckberry Nature 2023"
      )
  ) %>%
  pivot_longer(c(tpm, count), names_to = "count_type", values_to = "value") %>%
  mutate(
    is_zero = value == 0,
    value = value + 0.01
  ) %>%
  ggplot(
    aes(
      x = transcript_id,
      y = value,
      fill = treatment,
      shape = is_zero
    )
  ) +
    geom_quasirandom(color = alpha("black", 0)) +
    scale_y_log10(labels = scales::trans_format("log10", scales::math_format(10^.x))) +
    # triangle down for zero values, circle for non-zero
    scale_shape_manual(values = c(`FALSE` = 21, `TRUE` = 25), guide = "none") +
    theme_light() +
    guides(color = "none", fill = "none") +
    facet_grid(vars(count_type), vars(treatment), scales = "free_y") +
    labs(x = NULL) +
    theme(
      axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1)
    )

dir.create("plots")
ggsave(
  here("plots", "sendai_buckberry_by_treatment_tpm_beeswarm.pdf"),
  p,
  width = 7.5, height = 3.5
)

```


```{r}
liu_timepoints <- c(
  "D3", "D7", "D10", "D13", "D21", "P3", "P10"
)

liu_2020_sendai_tpm <- sendai_tpm %>%
  inner_join(
    sample_meta %>%
      filter(
        study_name == "Liu Nature 2020"
      )
  ) %>%
  pivot_longer(c(tpm, count), names_to = "count_type", values_to = "value") %>%
  mutate(
    is_zero = value == 0,
    value = value + 0.01,
    time_point = factor(time_point, levels = liu_timepoints, ordered = TRUE)
  )

p <- liu_2020_sendai_tpm %>%
  filter(count_type == "tpm") %>%
  ggplot(
    aes(
      x = time_point,
      y = value,
      fill = transcript_id,
      shape = is_zero
    )
  ) +
    geom_quasirandom(color = alpha("black", 0)) +
    scale_y_log10(labels = scales::trans_format("log10", scales::label_math(10^.x))) +
    # triangle down for zero values, circle for non-zero
    scale_shape_manual(values = c(`FALSE` = 21, `TRUE` = 25), guide = "none") +
    theme_light() +
    guides(color = "none", fill = guide_legend(override.aes = list(shape = 21))) +
    facet_wrap(vars(treatment), scales = "free_x")

dir.create("plots")
ggsave(
  here("plots", "sendai_liu_nature_2020_tpm_beeswarm_over_time.pdf"),
  p,
  width = 6, height = 3.5
)

```
