---
title: "Wrangle metadata"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library(here)
library(tidyverse)
library(powerjoin)
library(synExtra)

synapser::synLogin()
```

Metadata and reads already downloaded using `pysradb` and `fasterq-dump`.

## Metadata

```{r}
projects <- tribble(
  ~study_name, ~study_accession, ~geo_accession,
  "Liu Nature 2020", "SRP259918", "GSE149694",
  "Buckberry Nature 2023", "SRP286549", "GSE159114",
  "Liu Nat Met 2017", "SRP115256", NA,
  "Ji CSC 2016", "SRP059279", "GSE69692",
  "Bayerl CSC 2021", "SRP262162", "GSE150772",
  "Pastor CSC 2016", "SRP068579", "GSE76970"
)

run_metadata_fastqdl <- projects %>%
  mutate(
    data = map(
      study_accession,
      \(x) read_tsv(here("raw", x, "meta.tsv"), na = c("<NA>", "", "NA")) %>%
        select(-study_accession)
    )
  ) %>%
  unnest(data) %>%
  janitor::remove_empty("cols")

write_csv(
  run_metadata_fastqdl,
  here("raw", "run_meta", "run_metadata_fastqdl.csv")
)
```


```{r}
dir.create(here("raw", "run_meta"))
run_meta_xml <- projects %>%
  mutate(
    run_meta = map(
      study_accession,
      \(x) processx::run(
        "efetch",
        c(
          "-db", "sra", "-format", "native",
          "-id", x
        ),
      )
    )
  )
pwalk(
  run_meta_xml,
  \(run_meta, study_accession, ...) {
    write_file(
      run_meta$stdout,
      here("raw", "run_meta", paste0(study_accession, ".xml"))
    )
  }
)

```

```{r}
read_sra_xml <- function(file) {
  library(xml2)
  meta_xml <- read_xml(file)
  recs <- xml_find_all(meta_xml, "//RUN")
  extract_values <- function(n, tag) {
    nodes <- xml_find_all(n, tag)
    df <- map(nodes, xml_attrs) %>%
      map(as.data.frame.list) %>%
      data.table::rbindlist(fill = TRUE)
    df[["value"]] <- map_chr(nodes, xml_text)
    df
  }
  files <- map(recs, \(x) extract_values(x, ".//SRAFile"))
  primary_id <- map_chr(recs, \(x) xml_find_all(x, "./IDENTIFIERS/PRIMARY_ID") %>% xml_text())
  combined_meta <- tibble(
    run_accession = primary_id,
    files = files
  ) %>%
    unnest(files)
}

run_metadata_extracted <- projects %>%
  mutate(
    data = map(
      study_accession,
      \(x) read_sra_xml(here("raw", "run_meta", paste0(x, ".xml")))
    )
  ) %>%
  unnest(data) %>%
  janitor::remove_empty("cols")

write_csv(
  run_metadata_extracted,
  here("raw", "run_meta", "run_metadata_extracted.csv")
)
# run_metadata_extracted <- read_csv(
#   here("raw", "run_meta", "run_metadata_extracted.csv")
# )
```


```{r}

original_filenames <- run_metadata_extracted %>%
  filter(supertype == "Original") %>%
  select(run_accession, filename)

original_filenames_trimmed <- original_filenames %>%
  mutate(
    trimmed = str_replace(
      filename,
      "(_L[0-9]+|_lane_[0-9]+)?((_|\\.)R?[12])?(_001)?\\.fastq(\\.gz)?",
      ""
    )
  ) %>%
  distinct(run_accession, trimmed)
```


`run_metadata_fastqdl` contains some duplicate samples. Same biological sample
was run multiple times. Have to merge them later.

Ugh the metadata from SRA is missing some of the sample attributes. Have to
go to Biosample manually and get the xml files

https://www.ncbi.nlm.nih.gov/biosample?Db=biosample&DbFrom=bioproject&Cmd=Link&LinkName=bioproject_biosample&LinkReadableName=BioSample&ordinalpos=1&IdsFromResult=667717
https://www.ncbi.nlm.nih.gov/biosample?Db=biosample&DbFrom=bioproject&Cmd=Link&LinkName=bioproject_biosample&LinkReadableName=BioSample&ordinalpos=1&IdsFromResult=629766
https://www.ncbi.nlm.nih.gov/biosample?Db=biosample&DbFrom=bioproject&Cmd=Link&LinkName=bioproject_biosample&LinkReadableName=BioSample&ordinalpos=1&IdsFromResult=397941


```{r}
dir.create(here("raw", "sample_meta"))
sample_meta_xml <- run_metadata_fastqdl %>%
  distinct(study_accession, biosample) %>%
  group_nest(study_accession) %>%
  mutate(
    sample_meta = map(
      data,
      \(x) processx::run(
        "efetch",
        c(
          "-db", "biosample", "-format", "xml",
          "-id", paste(x$biosample, collapse = ",")
        ),
      )
    )
  )

pwalk(
  sample_meta_xml,
  \(sample_meta, study_accession, ...) {
    write_file(
      sample_meta$stdout,
      here("raw", "sample_meta", paste0(study_accession, ".xml"))
    )
  }
)
```


Buckberry samples are most likely mislabeled in SRA. Based on their filenames
we can try to reconstruct their actual identity.

The `title` column is completely wrong. Discarding it. `source_name` and `treatment`
are shifted.

The correct labels are manually curated from the file names.

```{r}
buckberry_manual_labels <- tibble::tribble(
  ~biosample, ~biosample_correct, ~time_point,
  "SAMN16380671", "SAMN16380671", "P13 + 20",
  "SAMN16380670", "SAMN16380670", "P13 + 14",
  "SAMN16380669", "SAMN16380669", "P15",
  "SAMN16380668", "SAMN16380668", "P17",
  "SAMN16380667", "SAMN16380667", "P18",
  "SAMN16380666", "SAMN16380666", "P17",
  "SAMN16380665", "SAMN16380665", "P16",
  "SAMN16380664", "SAMN16380664", "P8",
  "SAMN16380663", "SAMN16380662", "P20 + 3",
  "SAMN16380662", "SAMN16380661", "P4 + 19",
  "SAMN16380661", "SAMN16380656", "P9 + 13",
  "SAMN16380656", "SAMN16380663", "P15"
)
```

```{r}
read_biosample_xml <- function(file) {
  library(xml2)
  meta_xml <- read_xml(file)
  recs <- xml_find_all(meta_xml, "//BioSample")
  extract_values <- function(n, tag) {
    nodes <- xml_find_all(n, tag)
    df <- map(nodes, xml_attrs) %>%
      map(as.data.frame.list) %>%
      data.table::rbindlist(fill = TRUE)
    df[["value"]] <- map_chr(nodes, xml_text)
    df
  }
  ids <- map(recs, \(x) extract_values(x, ".//Id"))
  attributes <- map(recs, \(x) extract_values(x, ".//Attribute"))
  title <- map_chr(recs, \(x) xml_find_all(x, ".//Title") %>% xml_text())
  ids_wide <- ids %>%
    map(
      \(x) select(x, db, value) %>% pivot_wider(names_from = db, values_from = value)
    ) %>%
    bind_rows()
  combined_meta <- ids_wide %>%
    mutate(
      title = title,
      data = map(
        attributes,
        \(x) select(x, attribute_name, value) %>%
          pivot_wider(names_from = attribute_name, values_from = value)
      )
    ) %>%
    unnest(data)
}

biosample_meta_raw <- bind_rows(
  read_biosample_xml(here("raw", "sample_meta", "SRP286549.xml")) %>%
    transmute(
      biosample = BioSample, GEO, source_name,
      background, treatment,
    ) %>%
    inner_join(
      buckberry_manual_labels,
      by = "biosample"
    ) %>%
    mutate(
      treatment = treatment[match(biosample_correct, biosample)] %>%
        recode(`Control` = "Control ESC"),
      source_name = source_name[match(biosample_correct, biosample)]
    ) %>%
    select(-biosample_correct),
  read_biosample_xml(here("raw", "sample_meta", "SRP259918.xml")) %>%
    separate(
      `cell subtype/time point`,
      into = c("treatment", "time_point"),
      sep = "-"
    ) %>%
    transmute(
      biosample = BioSample, GEO, title, source_name,
      background = `source cell type`, cell_type = `cell type`,
      # treatment = recode(treatment, Fibroblast = "control"),
      treatment,
      time_point
    ),
  read_biosample_xml(here("raw", "sample_meta", "SRP115256.xml")) %>%
    transmute(
      biosample = BioSample, title, source_name = `NA`,
      tissue, age, isolate
    ) %>%
    extract(
      source_name,
      into = c("time_point", NA, "treatment"),
      regex = "(.*) ([0-9]+[FM]+|H9) (.*)",
      remove = FALSE
    ) %>%
    mutate(
      treatment = str_replace(treatment, " replicate [12]", "") %>%
        str_replace("^r", ""),
      time_point = str_replace(time_point, "plus", "+")
    ),
  read_biosample_xml(here("raw", "sample_meta", "SRP059279.xml")) %>%
    transmute(
      biosample = BioSample, title, source_name,
      cell_type = `cell type`
    ),
  read_biosample_xml(here("raw", "sample_meta", "SRP262162.xml")) %>%
    transmute(
      biosample = BioSample, title, source_name,
      background = `cell line`, treatment = `culture conditions`, batch
    ),
  read_biosample_xml(here("raw", "sample_meta", "SRP068579.xml")) %>%
    filter(`library type` == "Ovation RNASeq System V2") %>%
    transmute(
      biosample = BioSample, title, source_name,
      time_point = paste0("P", str_replace(`passage number`, "Passage ", ""))
    )
)
```

SRR3106350 doesn't have original filename available.

For Pastor CSC 2016 we only use RNA-seq samples, so warnings here are fine.

```{r}
biosample_meta <- biosample_meta_raw %>%
  power_left_join(
    run_metadata_fastqdl %>%
      power_left_join(
        original_filenames_trimmed,
        by = c("run_accession"),
        check = check_specs(
          unmatched_keys_left = "warn",
          unmatched_keys_right = "warn",
          duplicate_keys_right = "warn",
          duplicate_keys_left = "warn"
        )
      ) %>%
      group_by(
        sample_accession, biosample, study_accession, study_name, geo_accession
      ) %>%
      summarize(
        filename_prefix = str_c(unique(trimmed), collapse = ","),
        .groups = "drop"
      ),
    by = c("biosample"),
    check = check_specs(
      unmatched_keys_left = "warn",
      unmatched_keys_right = "warn",
      duplicate_keys_right = "warn",
      duplicate_keys_left = "warn"
    )
  ) %>%
  select(
    study_name, study_accession, geo_accession,
    sample_accession, biosample, title, source_name,
    filename_prefix,
    everything(), -GEO
  )

write_csv(
  biosample_meta,
  here("raw", "biosample_meta.csv")
)

```

```{r}
meta_dir_syn <- synMkdir(
  "syn53061839", "meta"
)

synStoreMany(
  c(
    here("raw", "run_meta", "run_metadata_fastqdl.csv"),
    here("raw", "run_meta", "run_metadata_extracted.csv"),
    here("raw", "biosample_meta.csv")
  ),
  parentId = meta_dir_syn,
  forceVersion = FALSE
)

```

